src/ParserCombinators/Datatypes.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTSyntax #-}
+{-# LANGUAGE InstanceSigs #-}
+-- {-# LANGUAGE FlexibleInstances #-}
+-- {-# LANGUAGE UndecidableInstances #-}
+
+module ParserCombinators.Datatypes where
+
+import Control.Applicative (Alternative (..))
+import Control.Monad (MonadPlus)
+
+------------------------
+---- ParseErrorType ----
+------------------------
+
+data ParseErrorType i e
+  = EndOfInput
+  | Unexpected i
+  | Expected i i
+  | ExpectedEndOfFile i
+  | CustomError e
+  | Empty
```

Comment:
This seems fishy, when do you need `Empty`?

Reply:
It's for the Applicative instance

Reply:
This reminds me - you should test the lawfulness of your instances.

For example, you can try https://hackage.haskell.org/package/quickcheck-classes-0.6.5.0/docs/Test-QuickCheck-Classes.html

Reply:
Yeah, I'll make sure to do that